pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionMetaDataManager.java

```
@@ -174,89 +181,124 @@ public synchronized boolean containsFunction(String tenant, String namespace, St
+     * @param functionMetaData The function metadata in question
+     * @param delete Is this a delete operation
+     * @throws IllegalStateException if we are not the leader
+     * @throws IllegalArgumentException if the request is out of date.
      */
-    public synchronized CompletableFuture<RequestResult> updateFunction(FunctionMetaData functionMetaData) {
-
-        FunctionMetaData existingFunctionMetadata = null;
-        if (containsFunction(functionMetaData.getFunctionDetails().getTenant(),
-                functionMetaData.getFunctionDetails().getNamespace(),
-                functionMetaData.getFunctionDetails().getName())) {
-            existingFunctionMetadata = getFunctionMetaData(functionMetaData.getFunctionDetails().getTenant(),
-                    functionMetaData.getFunctionDetails().getNamespace(),
-                    functionMetaData.getFunctionDetails().getName());
+    public synchronized void updateFunctionOnLeader(FunctionMetaData functionMetaData, boolean delete)
+            throws IllegalStateException, IllegalArgumentException {
+        if (exclusiveLeaderProducer == null) {
+            throw new IllegalStateException("Not the leader");
+        }
+        boolean needsScheduling;
+        if (delete) {
+            needsScheduling = proccessDeregister(functionMetaData);
+        } else {
+            needsScheduling = processUpdate(functionMetaData);
+        }
+        Request.ServiceRequest serviceRequest = Request.ServiceRequest.newBuilder()
+                .setServiceRequestType(delete ? Request.ServiceRequest.ServiceRequestType.DELETE : Request.ServiceRequest.ServiceRequestType.UPDATE)
+                .setFunctionMetaData(functionMetaData)
+                .setWorkerId(workerConfig.getWorkerId())
+                .setRequestId(UUID.randomUUID().toString())
+                .build();
+        try {
+            lastMessageSeen = exclusiveLeaderProducer.send(serviceRequest.toByteArray());
+        } catch (Exception e) {
```

Shouldn't we return a 500 error to the end user?  If we call just "errorNotifier.triggerError(e)", the worker die and the end user will likely not get a response or a timeout error

---

good point. Changed

---

"errorNotifier.triggerError(e);" is still being called.  The worker might exist before exception gets bubbled up and a response send back

---

@srkukarni ^^^

---

So the question here is whats the right thing to do. If we are having issues to write into the producer, should the leader just reject the request saying Internal server error and hope that things will be better next time? Or is the right approach to trigger worker death?

---

We should return the error to the worker making the call to the leader, otherwise the worker might have to wait for a timeout.  I think we should just return an error and the user can retry.  There is no guarantee that restarting the worker or electing another leader will help solve the issue since all the workers have the same configuration.